### PR TITLE
feat(kuma-cp) delete related mesh resources with DeleteAll on ResourceManager

### LINKED ISSUE #191: Deleting a Mesh should delete all the entities associated to the Mesh
And prevent data-planes from executing requests. Useful for development.

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Summary

Delete related mesh resources (including secrets) when mesh entity is deleted.
Rework of #312 due to https://github.com/Kong/kuma/pull/312#discussion_r332279987

### Full changelog

* Implement DeleteAll on ResourceManager

### Issues resolved

Fix #191

### CODE REVIEW

File: pkg/core/resources/manager/manager.go [hunk 1]
```diff
@@ -62,6 +66,26 @@ func (r *resourcesManager) Delete(ctx context.Context, resource model.Resource,
 	return r.Store.Delete(ctx, resource, fs...)
 }
 
+func (r *resourcesManager) DeleteAll(ctx context.Context, fs ...store.DeleteAllOptionsFunc) error {
+	opts := store.NewDeleteAllOptions(fs...)
+
+	for _, typ := range r.Registry.ListTypes() {
```
Comment (yskopets) [inline pkg/core/resources/manager/manager.go]: I don't agree that logic that goes through the list of all supported types and calls `DeleteAll` on each of them should be a part of generic `ResourcesManager`.

By design, `ResourceStore` was a store for a single resource type. `ResourceManager` was introduced to complement `ResourceStore` with type-specific business logic.

In a generic case, business logic of `DeleteAll` is to delete all entities of a single resource type. It doesn't make sense to delete objects of all types when `DeleteAll` is called for `TrafficLog` or `TrafficPermission`.

On the other hand, delete objects of all types is business logic specific to `Mesh`. So, I would suggest to move that code into `MeshManager`

File: pkg/core/resources/manager/customizable_manager.go [hunk 2]
```diff
@@ -35,6 +38,15 @@ func (m *customizableResourceManager) Delete(ctx context.Context, resource model
 	return m.resourceManager(resource.GetType()).Delete(ctx, resource, fs...)
 }
 
+func (m *customizableResourceManager) DeleteAll(ctx context.Context, fs ...store.DeleteAllOptionsFunc) error {
+	for _, typ := range m.registry.ListTypes() {
```
Comment (yskopets) [inline pkg/core/resources/manager/customizable_manager.go]: Similarly to the comment in `ResourceManager`, I think, business logic that goes through the list of all supported types and calls `DeleteAll` on each if them is specific to `Mesh`.

`customizableResourceManager` is just a dispatcher. I would expect `DeleteAll` to be dispatched to a single type-specific `ResourceManager` (like it happens for other `CRUDL` operations)

File: pkg/core/managers/apis/mesh/mesh_manager.go [hunk 3]
```diff
@@ -82,12 +92,32 @@ func (m *meshManager) Delete(ctx context.Context, resource core_model.Resource,
 	// delete Mesh first to avoid a state where a Mesh could exist without a Built-in CA.
 	// even if removal of Built-in CA fails later on, delete opration can be safely tried again.
 	if err := m.store.Delete(ctx, mesh, fs...); err != nil {
+		if !core_store.IsResourceNotFound(err) { // ignore other errors so we can retry removing other resources
+			return err
+		}
+	}
+	opts := core_store.NewDeleteOptions(fs...)
+	// delete all the secrets including CA
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/core/managers/apis/mesh/mesh_manager.go]: ok, changed

File: pkg/core/resources/manager/manager.go [hunk 4]
```diff
@@ -62,6 +63,20 @@ func (r *resourcesManager) Delete(ctx context.Context, resource model.Resource,
...
 
+func (r *resourcesManager) DeleteAll(ctx context.Context, list model.ResourceList, fs ...store.DeleteAllOptionsFunc) error {
+	opts := store.NewDeleteAllOptions(fs...)
+
+	if err := r.List(ctx, list, store.ListByMesh(opts.Mesh)); err != nil {
+		return err
+	}
+	for _, obj := range list.GetItems() {
+		if err := r.Delete(ctx, obj, store.DeleteBy(model.MetaToResourceKey(obj.GetMeta()))); err != nil {
+			return err
```
Comment (yskopets) [inline pkg/core/resources/manager/manager.go]: Let's ignore `core_store.IsResourceNotFound(err)`

File: pkg/core/resources/manager/manager_test.go [hunk 5]
```diff
@@ -57,10 +57,38 @@ var _ = Describe("Resource Manager", func() {
...
+
+			// and resource from mesh-1 is deleted
+			res1 := sample.TrafficRouteResource{}
+			err = resManager.Get(context.Background(), &res1, store.GetByKey("default", "tr-1", "mesh-1"))
+			Expect(store.IsResourceNotFound(err)).To(BeTrue())
+
+			// and resource from mesh-2 is retained
+			res2 := sample.TrafficRouteResource{}
+			err = resManager.Get(context.Background(), &res2, store.GetByKey("default", "tr-1", "mesh-2"))
+			Expect(err).ToNot(HaveOccurred())
```
Comment (yskopets) [inline pkg/core/resources/manager/manager_test.go]: We should validate here that only `TrafficRoute` resources get removed

File: pkg/core/secrets/manager/manager.go [hunk 6]
```diff
@@ -68,6 +70,20 @@ func (s *secretManager) Delete(ctx context.Context, secret *secret_model.SecretR
...
 
+func (s *secretManager) DeleteAll(ctx context.Context, fs ...core_store.DeleteAllOptionsFunc) error {
+	list := &secret_model.SecretResourceList{}
+	opts := core_store.NewDeleteAllOptions(fs...)
+	if err := s.secretStore.List(context.Background(), list, core_store.ListByMesh(opts.Mesh)); err != nil {
+		return err
+	}
+	for _, item := range list.Items {
+		if err := s.Delete(ctx, item, core_store.DeleteBy(model.MetaToResourceKey(item.Meta))); err != nil {
+			return err
```
Comment (yskopets) [inline pkg/core/secrets/manager/manager.go]: Let's ignore `core_store.IsResourceNotFound(err)`

File: pkg/core/managers/apis/mesh/mesh_manager.go [hunk 7]
```diff
@@ -82,13 +96,47 @@ func (m *meshManager) Delete(ctx context.Context, resource core_model.Resource,
...
+	for _, typ := range m.registry.ListTypes() {
+		list, err := m.registry.NewList(typ)
+		if err != nil {
+			return err
+		}
+		if err := m.otherManagers.DeleteAll(ctx, list, core_store.DeleteAllByMesh(opts.Name)); err != nil {
+			return errors.Wrap(err, "could not delete associated resources")
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply pkg/core/managers/apis/mesh/mesh_manager.go]: Makes sense, if we want to have consistent behaviour across managers, we need to do it.

**Pull request merged**